### ISSUE TITLE
Best option for teaching manual memory

### ISSUE THREAD

**Issue by waciumawanjohi** [ossu/computer-science#516]
CS 50 is now the first course in core systems. For years it was the only online course that taught and assessed manual memory (a topic particularly well suited to exploration in the C language).

The course is no longer alone, there are many offerings on this topic. Which should OSSU recommend?

* [UofHelsinki](http://moocfi.github.io/courses/2016/aalto-c/en/)

I took this course before hearing of CS50. I found it comprehensive and all that I needed to succeed. The multiple programming assignments were graded locally (on my machine) and remotely. The remote option is no longer available, but i believe the local option still works. (If this is not the case, the course does not belong in this list)

* [Duke](https://www.coursera.org/learn/interacting-system-managing-memory)

This course on Managing Memory is part of a [4 course series](https://www.coursera.org/specializations/c-programming) introducing programming with C.

* [Dartmouth](https://www.edx.org/course/c-programming-pointers-and-memory-management)

Similarly, this is the 4th course in a 7 course certificate in [C Programming with Linux](https://www.edx.org/professional-certificate/dartmouth-imtx-c-programming-with-linux).

* [Text Resource](http://cslibrary.stanford.edu/102/PointersAndMemory.pdf)

This primer on manual memory in C may be appropriate for the Readings page in extras.

* [CS50](https://www.edx.org/course/cs50s-introduction-computer-science-harvardx-cs50x)

A favorite of many.

Please share your thoughts. Particularly useful will be evaluations from students who have explored the newer options.

**Comment by hanjiexi** [comment]
The Helsinki one looks nice. Unlike Dart and Dartmouth, it is self-contained and it quickly gets you to dynamic memory allocation. The Dart and Dartmouth ones assume you've already done previous courses in their sequence so that would be a problem for anyone who doesn't have C experience.

The Helsinki course also seems to have a lot of mini exercises, which is great. On the other hand, it doesn't seem to take you as far as CS50 in terms of implementing some sort of (relatively) challenging problem that involves explicit memory management. If I remember correctly from my time in CS50, there was some project involving a long list of words and I had to implement a trie to rapidly traverse the data. Were any of the Helsinki course projects as challenging as the final C project in CS50?

**Comment by hanjiexi** [comment]
Now that we're using CS50 mainly as a pre-requisite for systems-related courses requiring knowledge of C, I think I do want to replace CS50 with Helsinki / [Aalto C](http://moocfi.github.io/courses/2016/aalto-c/en/). Before CS50 gets to C there is all that intro-to-CS stuff, like programming in Scratch, that we're handling elsewhere. The Helsinki course will be more efficient.

Losing CS50 does make me sad as it's such an inspiration to many students, including myself. Perhaps it could be listed as an optional alternative to the MIT Intro to CS.

**Comment by waciumawanjohi** [comment]
On CS50's place as an inspiration to students, I wonder if we shouldn't just put a thank you article in the Extras. We can talk about how CS50 was one of the earliest MOOCs and continues to forge new ground in CS education. We can also talk about the reasons that it doesn't quite fit the OSSU CS curriculum. Such an article can both serve the nostalgia of former students, peak the interest of new students and assuage students who think CS50 may not have been considered.

**Comment by KathrynN** [comment]
We could also perhaps point more advanced students towards CS50 instead in that part?  CS50 was probably one of the best introductions to Computer Science _for me_ because I was a self-taught programmer.  I knew all the programming stuff, the loops and functions and even http protocols, but I'd never really had the theory behind it before.  Or the memory management, thank you garbage collection!

**Comment by hanjiexi** [comment]
@KathrynN I agree with your words on CS50's value to those who already have some experience and need a greater challenge to inspire them.

> We could also perhaps point more advanced students towards CS50 instead in that part?

Do you mean in Intro CS or the current place where CS50 is now (as the first course in Systems)? In either case, how the new version of the curriculum will end up being structured is still in flux. We might end up with a more flexible choose-your-guided-path curriculum or with the more rigid one we have now. I am not sure what's better for the majority of students.

**Comment by doct0rX** [comment]
i'd like to keep the curriculum as it was (moving CS50 at the top back again). As i took some programming courses before starting OSSU and starting CS50 i wasn't having any prior experience in C then I started to go with the course and tackle down it's challenges by further searching and reading some blogs and watching youtube videos to get the content and even further more and i was really enjoying it.
So i think we can keep it with a some prior C course for those who are totally new to CS field.

**Comment by TheNoLifeKing** [comment]
I'm in favor of leaving CS50 at the top as well. (In introduction). It was, for me, the perfect introduction to computer science. Perhaps there's some discussion worth having when to stop CS50, but with all the CS content I've ever digested, this was absolutely the best Introduction to Computer Science and a perfect tool for learning if CS is right for you or not (Which is the point of these introduction courses, is it not?

**Comment by waciumawanjohi** [comment]
The auto grader for the University of Helsinki course no longer works. I have been working on reaching out to the course owner. Until that course is available, CS50 remains the best alternative.
